File: backend/app/metrics.py
import time
from typing import Dict, List
from collections import defaultdict

class BackendMetrics:
    auth_failures: int = 0
    validation_failures: int = 0
    duplicate_requests: int = 0
    provider_failures: Dict[str, int] = defaultdict(int)
    api_latencies: Dict[str, List[float]] = defaultdict(list)

    @classmethod
    def record_auth_failure(cls):
        cls.auth_failures += 1

    @classmethod
    def record_validation_failure(cls):
        cls.validation_failures += 1

    @classmethod
    def record_duplicate_request(cls):
        cls.duplicate_requests += 1

    @classmethod
    def record_provider_failure(cls, provider: str):
        cls.provider_failures[provider] += 1

    @classmethod
    def record_latency(cls, path: str, latency_ms: float):
        # Keep only the last 100 latencies per path to avoid memory growth
        latencies = cls.api_latencies[path]
        latencies.append(latency_ms)
        if len(latencies) > 100:
            latencies.pop(0)

    @classmethod
    def get_summary(cls) -> dict:
        avg_latencies = {}
        for path, vals in cls.api_latencies.items():
            if vals:
                avg_latencies[path] = sum(vals) / len(vals)
        return {
            "auth_failures": cls.auth_failures,
            "validation_failures": cls.validation_failures,
            "duplicate_requests": cls.duplicate_requests,
            "provider_failures": dict(cls.provider_failures),
            "average_latencies_ms": avg_latencies,
        }

File: frontend/lib/metrics.ts
export interface OperationalMetrics {
  authFailures: number;
  validationFailures: number;
  duplicateRequests: number;
  networkErrors: number;
  latencies: Record<string, number[]>;
}

export const metrics: OperationalMetrics = {
  authFailures: 0,
  validationFailures: 0,
  duplicateRequests: 0,
  networkErrors: 0,
  latencies: {},
};

export const recordAuthFailure = () => {
  metrics.authFailures++;
};

export const recordValidationFailure = () => {
  metrics.validationFailures++;
};

export const recordDuplicateRequest = () => {
  metrics.duplicateRequests++;
};

export const recordNetworkError = () => {
  metrics.networkErrors++;
};

export const recordLatency = (path: string, latencyMs: number) => {
  if (!metrics.latencies[path]) {
    metrics.latencies[path] = [];
  }
  const pathLatencies = metrics.latencies[path];
  pathLatencies.push(latencyMs);
  if (pathLatencies.length > 50) {
    pathLatencies.shift();
  }
};

if (typeof window !== "undefined") {
  (window as any).__repodar_metrics = {
    getSummary: () => {
      const avgLatencies: Record<string, string> = {};
      for (const [path, list] of Object.entries(metrics.latencies)) {
        if (list.length > 0) {
          const sum = list.reduce((a, b) => a + b, 0);
          avgLatencies[path] = `${(sum / list.length).toFixed(1)}ms`;
        }
      }
      return {
        authFailures: metrics.authFailures,
        validationFailures: metrics.validationFailures,
        duplicateRequests: metrics.duplicateRequests,
        networkErrors: metrics.networkErrors,
        averageLatencies: avgLatencies,
      };
    },
    raw: metrics,
  };
}
